### ISSUE TITLE
[keymgr] Security enhancement

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#8120]
- for the random values used to wipe keys in key manager, we can consider additional permutation on one of the shares to ensure the values are diverse.

**Comment by msfschaffner** [comment]
I wasn't sure whether this is supposed to be an enhancement, or something that we need to do - hence I labeled for D2S.
Please relabel if this is inaccurate.

**Comment by tjaychen** [comment]
yeah this is meant to be marked as a future enhancement, since i didn't want to alter the existing keymgr more.

**Comment by ballifatih** [comment]
Triaged for `keymgr`. It is unclear what could go wrong if we do not implement randomization. Temporarily marking this as `Type:FutureRelease`.

**Comment by ballifatih** [comment]
For the first case (inside `keymgr_ctrl`), I am not sure I understood the question exactly. Are you suggesting that we have a secret netlist constant value, and reset the value of state_d to this fixed value instead of '0 vector? The reset value of state_d (which is effectively '0) lingers until the first advance command is executed (which latches the OTP root key).

Otherwise, as you also pointed out, because of the XORing behavior, we need to load equal-share randomness. If we do not load equal-value shares, then after XORing, we would not get the correct key value (i.e. lose the value of OTP root key).  Therefore the effective key value of pre-populated randomness must be '0. This value lingers in state_d for few clock cycles (8 clock cycles, if EDN/LFSR is not blocking), so I do not think it would be a big issue. No SW interaction is required for switching from this randomness to OTP-loaded state_d value.

For the second one (inside `keymgr_kmac_if`), I think this value is likely to be propagated to KMAC as the key value. I agree that it would be better if the effective key behind the two shares is also random. The easiest solution is to use `entropy_i[0]` and `entropy_i[1]` respectively for two shares. This is still not a very good solution because the same randomness `entropy_i[1]` would be shared for two different purposes.

Another point about these relatively small entropy is that if KMAC can be invoked with these random keys, it would leak the state of internal LFSR to some point. 32-bit space is small enough that we can brute-force through it.

**Comment by vogelpi** [comment]
> For the first case (inside `keymgr_ctrl`), I am not sure I understood the question exactly. Are you suggesting that we have a secret netlist constant value, and reset the value of state_d to this fixed value instead of '0 vector? The reset value of state_d (which is effectively '0) lingers until the first advance command is executed (which latches the OTP root key).
> 
> Otherwise, as you also pointed out, because of the XORing behavior, we need to load equal-share randomness. If we do not load equal-value shares, then after XORing, we would not get the correct key value (i.e. lose the value of OTP root key). Therefore the effective key value of pre-populated randomness must be '0. This value lingers in state_d for few clock cycles (8 clock cycles, if EDN/LFSR is not blocking), so I do not think it would be a big issue. No SW interaction is required for switching from this randomness to OTP-loaded state_d value.

This is exactly the kind of information I was looking for! This means the effective 0 value won't be in the register for long. I think we don't need to change this then.

> For the second one (inside `keymgr_kmac_if`), I think this value is likely to be propagated to KMAC as the key value. I agree that it would be better if the effective key behind the two shares is also random. The easiest solution is to use `entropy_i[0]` and `entropy_i[1]` respectively for two shares. This is still not a very good solution because the same randomness `entropy_i[1]` would be shared for two different purposes.

Yes, that's true. If we changed his `entropy_i[0] ^ entropy_i[1]` would be used as key and `entropy_i[1]` as data input for KMAC. But I think it's probably still better than forwarding effective 0 as sideload key to KMAC, AES and OTBN by default (the  `data_o` output of `keymgr_kmac_if` is forwarded to the `key_i` inputs of the sideload ports inside `keymgr_sideload_key_ctrl.sv`.

> Another point about these relatively small entropy is that if KMAC can be invoked with these random keys, it would leak the state of internal LFSR to some point. 32-bit space is small enough that we can brute-force through it.

I think it's true that there is relatively little entropy here. However, the LFSR is 64-bit so I am not too worried about it. We are using individual 32-bit chunks yes, but to know future and past 32-bit chunks you need to know the entire 64-bit LFSR which is not brute-forcable.

To summarize, I think we should change `keymgr_kmac_if.sv` to no forward an effective 0 key by default. Everything else should be fine. Thanks a lot @ballifatih for your input!

**Comment by vogelpi** [comment]
Earlgrey-PROD.M3 and P2 sounds right to me here.
